p2p/src/transport.rs

```diff
@@ -0,0 +1,22 @@
+#[derive(Debug, thiserror::Error)]
+pub enum Error {}
+
+pub trait Connection: Clone {
+    type Stream;
+
+    fn open(&self) -> Result<Self::Stream, Error>;
+}
+
+pub trait Endpoint {
+    type Connection;
+
+    fn connect(&self) -> Result<Self::Connection, Error>;
+}
+
+pub trait Transport {
+    type Connection: Connection;
+    type Endpoint: Endpoint<Connection = <Self as Transport>::Connection>;
+    type Incoming: Iterator<Item = Result<<Self as Transport>::Connection, Error>>;
+
+    fn bind(&self) -> Result<(Self::Endpoint, Self::Incoming), Error>;
```

Comment:
is bind like a constructor? where it gets endpoint and incoming connections?

Reply:
It's a constructor that returns a working endpoint and the stream of incoming connections. Inspired by the syscall it is to indicate that after that call connections can be created and accepted.